pokemon.go

```diff
@@ -71,6 +130,115 @@ func (p *Pokemon) HasValidEVs() bool {
+func (p *Pokemon) GainExperience(exp int) {
+
+	if exp < 0 {
+		panic(fmt.Sprintf("%s's experience tried to decrease by %d", p.GetName(), exp))
+	}
+
+	if exp == 0 {
+		return
+	}
+
+	max_exp := EXP_TABLE[p.GetGrowthRate()][MAX_LEVEL]
+
+	// if would gain experience beyond leveling to 100, set level to 100
+	if int(p.TotalExperience)+exp > max_exp {
+		p.GainLevels(MAX_LEVEL - int(p.Level))
+		return
+	}
+
+	remaining_exp := exp
+	toNextLevel := EXP_TABLE[p.GetGrowthRate()][int(p.Level+1)] - int(p.TotalExperience)
+
+	for remaining_exp > toNextLevel {
+		p.GainLevels(1)
+		remaining_exp -= toNextLevel
+		toNextLevel = EXP_TABLE[p.GetGrowthRate()][int(p.Level+1)] - int(p.TotalExperience)
+	}
+
+	// add whats left
+	p.TotalExperience += uint(remaining_exp)
+
+}
+
+func (p *Pokemon) computeStats() {
+	if !p.HasValidLevel() {
+		panic(fmt.Sprintf("Failed to compute stats for %s, level %d invalid", p.GetName(), p.Level))
+	}
+
+	if !p.HasValidIVs() {
+		panic(fmt.Sprintf("Failed to compute stats for %s, ivs %d invalid", p.GetName(), p.IVs))
+	}
+
+	if !p.HasValidEVs() {
+		panic(fmt.Sprintf("Failed to compute stats for %s, evs %d invalid", p.GetName(), p.EVs))
+	}
+
+	// get base stats
+	base_stats := p.GetBaseStats()
+
+	// compute HP
+	hp_ev_term := math.Floor(float64(p.EVs[STAT_HP]) / 4)
+	base_iv_ev_level_hp_term := (2*uint(base_stats[STAT_HP]) + uint(p.IVs[STAT_HP]) + uint(hp_ev_term)) * uint(p.Level)
+	hp_floor_term := math.Floor(float64(base_iv_ev_level_hp_term) / 100)
+	p.Stats[STAT_HP] = uint(hp_floor_term + float64(p.Level) + 10)
+	p.CurrentHP = p.Stats[STAT_HP]
+
+	// compute all other stats
+	natureModifiers := p.Nature.getNatureModifers()
+	for _, stat := range [5]int{STAT_ATK, STAT_DEF, STAT_SPATK, STAT_SPDEF, STAT_SPD} {
+		ev_term := math.Floor(float64(p.EVs[stat]) / 4)
+		base_iv_ev_level_term := (2*uint(base_stats[stat]) + uint(p.IVs[stat]) + uint(ev_term)) * uint(p.Level)
+		floor_term := math.Floor(float64(base_iv_ev_level_term) / 100)
+		nature_term := float64(floor_term+5) * natureModifiers[stat]
+		p.Stats[stat] = uint(math.Floor(nature_term))
+	}
+}
+
+func (p *Pokemon) VerboseString() string {
+	outputString := ""
+	outputString += "{\n" +
+		fmt.Sprintf("\tNational dex number: %d\n", p.NatDex) +
+		fmt.Sprintf("\tLevel: %d\n", p.Level) +
+		fmt.Sprintf("\tAbility: %s\n", p.Ability) +
+		fmt.Sprintf("\tTotal experience: %d\n", p.TotalExperience) +
+		fmt.Sprintf("\tGender: %s\n", p.Gender) +
+		fmt.Sprintf("\tIVs: %v\n", p.IVs) +
+		fmt.Sprintf("\tEVs: %v\n", p.EVs) +
+		fmt.Sprintf("\tNature: %s\n", p.Nature) +
+		fmt.Sprintf("\tStats: %v\n", p.Stats) +
+		fmt.Sprintf("\tCurrent HP: %d\n", p.CurrentHP) +
+		fmt.Sprintf("\tHeld item: %v\n", p.HeldItem) +
+		fmt.Sprintf("\tMoves: %v\n", p.Moves) +
+		fmt.Sprintf("\tFriendship: %d\n", p.Friendship) +
+		fmt.Sprintf("\tOriginal trainer ID: %d\n", p.OriginalTrainerID) +
+		"}"
+	return outputString
+}
```

Comment:
What's this for? It's also not covered in tests

Reply:
this mainly exists for use in tests